File: src/main/java/com/dataloom/hazelcast/serializers/EntityStreamSerializer.java
package com.dataloom.hazelcast.serializers;

import java.io.IOException;
import java.util.Map;

import org.springframework.stereotype.Component;

import com.dataloom.data.EntityKey;
import com.dataloom.hazelcast.StreamSerializerTypeIds;
import com.dataloom.linking.Entity;
import com.hazelcast.nio.ObjectDataInput;
import com.hazelcast.nio.ObjectDataOutput;
import com.kryptnostic.rhizome.pods.hazelcast.SelfRegisteringStreamSerializer;

@Component
public class EntityStreamSerializer implements SelfRegisteringStreamSerializer<Entity> {

    @Override
    public void write( ObjectDataOutput out, Entity object ) throws IOException {
        EntityKeyStreamSerializer.serialize( out, object.getKey() );
        //TODO Warning: write a map stream serializer?
        out.writeObject( object.getProperties() );
    }

    @Override
    public Entity read( ObjectDataInput in ) throws IOException {
        EntityKey ek = EntityKeyStreamSerializer.deserialize( in );
        //TODO Warning: write a map stream serializer?
        Map<String, Object> properties = in.readObject();
        return new Entity( ek, properties );
    }

    @Override
    public int getTypeId() {
        return StreamSerializerTypeIds.ENTITY.ordinal();
    }

    @Override
    public void destroy() {}

    @Override
    public Class<? extends Entity> getClazz() {
        return Entity.class;
    }

}

File: src/main/kotlin/com/openlattice/hazelcast/serializers/EntityStreamSerializer.kt
package com.openlattice.hazelcast.serializers

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.hazelcast.nio.ObjectDataInput
import com.hazelcast.nio.ObjectDataOutput
import com.openlattice.IdConstants
import com.openlattice.IdConstants.LAST_WRITE_ID
import com.openlattice.IdConstants.VERSION_ID
import com.openlattice.data.Entity
import com.openlattice.hazelcast.StreamSerializerTypeIds
import com.openlattice.mapstores.TestDataFactory
import org.apache.olingo.commons.api.edm.EdmPrimitiveTypeKind
import org.springframework.stereotype.Component
import java.time.OffsetDateTime
import javax.inject.Inject

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
@Component
class EntityStreamSerializer : TestableSelfRegisteringStreamSerializer<Entity> {
    @Inject
    private lateinit var mapper: ObjectMapper

    override fun generateTestValue(): Entity = Entity(
            TestDataFactory.entities(1,
                                     setOf(
                                             TestDataFactory.propertyType(
                                                     EdmPrimitiveTypeKind.String
                                             ),
                                             TestDataFactory.propertyType(EdmPrimitiveTypeKind.String)
                                     ).associateBy { it.id }).values.first()
    )

    override fun getTypeId(): Int = StreamSerializerTypeIds.ENTITY.ordinal

    override fun getClazz(): Class<out Entity> = Entity::class.java

    override fun write(out: ObjectDataOutput, obj: Entity) {
        out.writeByteArray(mapper.writeValueAsBytes(obj.properties))
    }

    override fun read(input: ObjectDataInput): Entity = Entity(mapper.readValue(input.readByteArray()))

}